snapshots/devmapper/config.go

```diff
@@ -0,0 +1,160 @@
+// +build linux
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package devmapper
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/BurntSushi/toml"
+	"github.com/containerd/containerd/snapshots/devmapper/dmsetup"
+	"github.com/docker/go-units"
+	"github.com/hashicorp/go-multierror"
+	"github.com/pkg/errors"
+)
+
+const (
+	// See https://www.kernel.org/doc/Documentation/device-mapper/thin-provisioning.txt for details
+	dataBlockMinSize = 128
+	dataBlockMaxSize = 2097152
+)
+
+var (
+	errInvalidBlockSize      = errors.Errorf("block size should be between %d and %d", dataBlockMinSize, dataBlockMaxSize)
+	errInvalidBlockAlignment = errors.Errorf("block size should be multiple of %d sectors", dataBlockMinSize)
+)
+
+// Config represents device mapper configuration loaded from file.
+// Size units can be specified in human-readable string format (like "32KIB", "32GB", "32Tb")
+type Config struct {
+	// Device snapshotter root directory for metadata
+	RootPath string `toml:"root_path"`
+
+	// Name for 'thin-pool' device to be used by snapshotter (without /dev/mapper/ prefix)
+	PoolName string `toml:"pool_name"`
+
+	// Path to data volume to be used by thin-pool
+	DataDevice string `toml:"data_device"`
+
+	// Path to metadata volume to be used by thin-pool
+	MetadataDevice string `toml:"meta_device"`
+
+	// The size of allocation chunks in data file.
+	// Must be between 128 sectors (64KB) and 2097152 sectors (1GB) and a multiple of 128 sectors (64KB)
+	// Block size can't be changed after pool created.
+	// See https://www.kernel.org/doc/Documentation/device-mapper/thin-provisioning.txt
+	DataBlockSize        string `toml:"data_block_size"`
+	DataBlockSizeSectors uint32 `toml:"-"`
+
+	// Defines how much space to allocate when creating base image for container
+	BaseImageSize      string `toml:"base_image_size"`
+	BaseImageSizeBytes uint64 `toml:"-"`
+}
+
+// LoadConfig reads devmapper configuration file from disk in TOML format
+func LoadConfig(path string) (*Config, error) {
+	if _, err := os.Stat(path); err != nil {
+		if os.IsNotExist(err) {
+			return nil, os.ErrNotExist
+		}
+
+		return nil, err
+	}
+
+	config := Config{}
```

Comment:
There is no way we can set default parameters and make it work without setting up thin-pool in advance. So if we set default params - user will be seeing errors in containerd logs like `failed to create thin-pool with name "containerd-pool": failed to get block device size: blockdev: cannot open: No such file or directory"` (which is confusing if user is not going to use devmapper) instead of [current](https://github.com/containerd/containerd/blob/14d78592169db3f3b3fed641d854f31a940193d3/snapshots/devmapper/snapshotter.go#L58) `device mapper not configured`, underlying that it needs more attention to make it work.

Reply:
make senses to me.